src/classes/AN_AutoNumber_TDTM.cls

```diff
@@ -0,0 +1,48 @@
+/*
+    Copyright (c) 2020 Salesforce.org
+    All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of Salesforce.org nor the names of
+      its contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2020
+* @description Trigger Handler for SObjects that have active AutoNumbers.
+*/
+public class AN_AutoNumber_TDTM extends TDTM_Runnable {
+
+    public override DmlWrapper run(
+            List<SObject> newList, List<SObject> oldList,
+            TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult
+    ) {
+        if (triggerAction == TDTM_Runnable.Action.AfterInsert) {
```

Comment:
Right.  It would make incrementing and keeping track of the current max used number more difficult if the code applied the batch number before insert and then there was a failure on the dml.  This ensures that only Batches(records) that are successfully inserted then get a Batch Number applied and the code can then update the records and increment the max used number with confidence.

Reply:
So, to follow up, this is the error message when mixed SObjects are passed in the list: 

```
System.SObjectException: DataImportBatch__c.Batch_Number__c does not belong to SObject type Account
Class.AN_AutoNumberService.apply: line 117, column 1
```